crates/ide_assists/src/handlers/extract_module.rs

```diff
@@ -0,0 +1,660 @@
+                            vis_node_occurences.push(it.syntax().clone());
+                        },
+                        ast::Fn(it) => {
+                            let node_def = {
+                                let nod = ctx.sema.to_def(&it)?;
+                                Definition::ModuleDef(nod.into())
+                            };
+
+                            self.get_and_process_usages(ctx, builder, node_def);
+                            vis_node_occurences.push(it.syntax().clone());
+                        },
+                        ast::Impl(it) => {
+                            for desc in it.syntax().descendants() {
+                                if let Some(fn_) = ast::Fn::cast(desc) {
+                                    vis_node_occurences.push(fn_.syntax().clone());
+                                }
+                            }
+                        },
+                        _ => (),
+                    }
+                };
+            }
+        }
+
+        let body = self.body.clone_for_update();
+
+        let mut replacements = Vec::new();
+        body.syntax().descendants().for_each(|desc| {
+            if let Some(_) = vis_node_occurences
+                .clone()
+                .into_iter()
+                .find(|x| x.kind() == desc.kind() && x.text() == desc.text())
+            {
+                match_ast! {
+                    match desc {
+                        ast::Struct(it) => replacements.push((it.visibility().clone(), it.syntax().clone())),
+                        ast::Union(it) => replacements.push((it.visibility().clone(), it.syntax().clone())),
+                        ast::Enum(it) => replacements.push((it.visibility().clone(), it.syntax().clone())),
+                        ast::TypeAlias(it) => replacements.push((it.visibility().clone(), it.syntax().clone())),
+                        ast::Const(it) => replacements.push((it.visibility().clone(), it.syntax().clone())),
+                        ast::Static(it) => replacements.push((it.visibility().clone(), it.syntax().clone())),
+                        ast::Fn(it) => replacements.push((it.visibility().clone(), it.syntax().clone())),
+                        ast::RecordField(it) => replacements.push((it.visibility().clone(), it.syntax().clone())),
+                        _ => (),
+                    }
+                }
+            }
+        });
+
+        replacements.into_iter().for_each(|(vis, syntax)| {
+            add_change_vis(vis, syntax.first_child_or_token());
+        });
+
+        Some(body)
+    }
+
+    fn get_and_process_usages(
+        &self,
+        ctx: &AssistContext,
+        builder: &mut AssistBuilder,
+        node_def: Definition,
+    ) {
+        for (file_id, references) in node_def.usages(&ctx.sema).all() {
+            let source_file = ctx.sema.parse(file_id);
+
+            builder.edit_file(file_id);
```

Comment:
This will be problematic, switching to the same file twice will only commit the changes once I believe and discard any previous changes. Instead you can collect the usages of everything in a map or something and do this for all references for each file together or something along those lines.